### ISSUE TITLE
Resizing extended partition fails (DOS style partition)

### ISSUE THREAD

**Issue by carlf** [karelzak/util-linux#520]
I have a machine with an extended partition containing a single LVM partition as shown here:

```
label: dos
label-id: 0x9781676e
device: /dev/sda
unit: sectors

/dev/sda1 : start=        2048, size=      583680, type=83, bootable
/dev/sda2 : start=      587774, size=   125239298, type=5
/dev/sda5 : start=      587776, size=   125239296, type=8e
```

When I try to extend partition 2 to fill the empty space, this is what I get:

```
root@av-measurement-db-psql-prod-sc-03:~# echo ',+' | LIBFDISK_DEBUG=all sfdisk -N2 --no-reread /dev/sda
2084: libfdisk:     INIT: library debug mask: 0xffff
Available "LIBFDISK_DEBUG=<name>[,...]|<mask>" debug masks:
   all      [0xffff] : info about all subsystems
   ask      [0x0010] : fdisk dialogs
   help     [0x0001] : this help
   cxt      [0x0004] : library context (handler)
   label    [0x0008] : disk label utils
   part     [0x0040] : partition utils
   parttype [0x0080] : partition type utils
   script   [0x0200] : sfdisk-like scripts
   tab      [0x0100] : table utils
2084: libfdisk:      CXT: [0x212ad50]: alloc
2084: libfdisk:      CXT: [0x212ad50]: assigning device /dev/sda
2084: libfdisk:      CXT: [0x212ad50]: *** resetting context
2084: libfdisk:      CXT: [0x212ad50]: /dev/sda: discovering topology...
2084: libfdisk:      CXT: [0x212ad50]: initialize libblkid prober
2084: libfdisk:      CXT: [0x212ad50]: result: log/phy sector size: 512/512
2084: libfdisk:      CXT: [0x212ad50]: result: fdisk/optimal/minimal io: 512/0/512
2084: libfdisk:      CXT: [0x212ad50]: /dev/sda: discovering geometry...
2084: libfdisk:      CXT: [0x212ad50]: total sectors: 314572800 (ioctl=314572800)
2084: libfdisk:      CXT: [0x212ad50]: result: C/H/S: 19581/255/63
2084: libfdisk:      CXT: [0x212ad50]: initialize in-memory first sector buffer [sector_size=512]
2084: libfdisk:      CXT: [0x212ad50]: reading: offset=0, size=512
2084: libfdisk:      CXT: [0x212ad50]: probing for gpt
2084: libfdisk:    LABEL: GPT probe failed
2084: libfdisk:      CXT: [0x212ad50]: probing for dos
2084: libfdisk:    LABEL: DOS: initialize, first sector buffer 0x212cd20
2084: libfdisk:    LABEL: DOS PT geometry: heads=0, sectors=0
2084: libfdisk:    LABEL: DOS: Reading extended 1
2084: libfdisk:    LABEL: DOS: reading EBR 4: offset=587774, buffer=0x2132270
2084: libfdisk:    LABEL: DOS: EBR[offset=587774]: link: type=0,  start=0, size=0;  data: type=8e, start=2, size=125239296
2084: libfdisk:    LABEL: DOS: nparts_max: 5
2084: libfdisk:      CXT: [0x212ad50]: --> switching context to dos!
2084: libfdisk:      CXT: [0x212ad50]: appling user device properties
2084: libfdisk:      CXT: [0x212ad50]: reseting alignment...
2084: libfdisk:      CXT: [0x212ad50]: alignment reseted to: first LBA=2048, last LBA=314572799, grain=1048576 [rc=0]
2084: libfdisk:      CXT: [0x212ad50]: new C/H/S: 19581/255/63
2084: libfdisk:      CXT: [0x212ad50]: new log/phy sector size: 512/512
2084: libfdisk:      CXT: [0x212ad50]: initialized for /dev/sda [READ-WRITE]
2084: libfdisk:   SCRIPT: [0x212cbf0]: alloc
2084: libfdisk:      TAB: [0x212b280]: alloc
2084: libfdisk:      CXT: [0x212ad50]: setting reference to script 0x212cbf0
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'
Disk /dev/sda: 150 GiB, 161061273600 bytes, 314572800 sectors
2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'
Units: sectors of 1 * 512 = 512 bytes
2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'
Sector size (logical/physical): 512 bytes / 512 bytes
2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'
I/O size (minimum/optimal): 512 bytes / 512 bytes
2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'
Disklabel type: dos
2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
2084: libfdisk:      CXT: [0x212ad50]: asking for disk dos ID
2084: libfdisk:      CXT: [0x212ad50]: asking for disk dos item 0
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'
Disk identifier: 0x9781676e
2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'

Old situation:
2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
2084: libfdisk:      CXT: [0x212ad50]:  -- get table --
2084: libfdisk:      TAB: [0x212d070]: alloc
2084: libfdisk:     PART: [0x2132900]: alloc
2084: libfdisk:      TAB: [0x212d070]: add entry 0x2132900 [start=2048, end=585727, size=583680,   primary]
2084: libfdisk:     PART: [0x21329e0]: alloc
2084: libfdisk:      TAB: [0x212d070]: add entry 0x21329e0 [start=587774, end=125827071, size=125239298,   container]
2084: libfdisk:     PART: [0x2132ac0]: alloc
2084: libfdisk:      TAB: [0x212d070]: add entry 0x2132ac0 [start=587776, end=125827071, size=125239296,  nested primary]
2084: libfdisk:      ASK: [0x212cf30]: alloc
2084: libfdisk:      ASK: [0x212cf30]: do_ask for 'info'

2084: libfdisk:      ASK: [0x212cf30]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x212cf30]: reset
2084: libfdisk:      ASK: [0x212cf30]: free
Device     Boot  Start       End   Sectors  Size Id Type
/dev/sda1  *      2048    585727    583680  285M 83 Linux
/dev/sda2       587774 125827071 125239298 59.7G  5 Extended
/dev/sda5       587776 125827071 125239296 59.7G 8e Linux LVM
2084: libfdisk:      TAB: [0x212d070]: wrong older check
2084: libfdisk:      TAB: [0x212d070]: reset
2084: libfdisk:      TAB: [0x212d070]: remove entry 0x2132900
2084: libfdisk:     PART: [0x2132900]: reset
2084: libfdisk:     PART: [0x2132900]: free
2084: libfdisk:      TAB: [0x212d070]: remove entry 0x21329e0
2084: libfdisk:     PART: [0x21329e0]: reset
2084: libfdisk:     PART: [0x21329e0]: free
2084: libfdisk:      TAB: [0x212d070]: remove entry 0x2132ac0
2084: libfdisk:     PART: [0x2132ac0]: reset
2084: libfdisk:     PART: [0x2132ac0]: free
2084: libfdisk:      TAB: [0x212d070]: free
2084: libfdisk:   SCRIPT: [0x212cbf0]: setting new header label='dos'

2084: libfdisk:   SCRIPT: [0x212cbf0]:  parsing line 0
2084: libfdisk:   SCRIPT: [0x212cbf0]:   parsing buffer
2084: libfdisk:     PART: [0x21329e0]: alloc
2084: libfdisk:   SCRIPT: [0x212cbf0]:  parsing item 0 (',+')
2084: libfdisk:   SCRIPT: [0x212cbf0]:  parsing item 1 ('')
2084: libfdisk:      TAB: [0x212b280]: add entry 0x21329e0 [start=18446744073709551615, end=18446744073709551613, size=18446744073709551615,   primary]
2084: libfdisk:     PART: [0x2132ac0]: alloc
2084: libfdisk:     PART: [0x21329e0]: >>> resize requested
2084: libfdisk:      CXT: [0x212ad50]:  -- get table --
2084: libfdisk:      TAB: [0x212d030]: alloc
2084: libfdisk:     PART: [0x2133a30]: alloc
2084: libfdisk:      TAB: [0x212d030]: add entry 0x2133a30 [start=2048, end=585727, size=583680,   primary]
2084: libfdisk:     PART: [0x2132890]: alloc
2084: libfdisk:      TAB: [0x212d030]: add entry 0x2132890 [start=587774, end=125827071, size=125239298,   container]
2084: libfdisk:     PART: [0x2132c60]: alloc
2084: libfdisk:      TAB: [0x212d030]: add entry 0x2132c60 [start=587776, end=125827071, size=125239296,  nested primary]
2084: libfdisk:      CXT: [0x212ad50]: -- get freespace --
2084: libfdisk:      CXT: [0x212ad50]:  -- get table --
2084: libfdisk:      TAB: [0x2132970]: alloc
2084: libfdisk:     PART: [0x2132d40]: alloc
2084: libfdisk:      TAB: [0x2132970]: add entry 0x2132d40 [start=2048, end=585727, size=583680,   primary]
2084: libfdisk:     PART: [0x2132e40]: alloc
2084: libfdisk:      TAB: [0x2132970]: add entry 0x2132e40 [start=587774, end=125827071, size=125239298,   container]
2084: libfdisk:     PART: [0x2132f20]: alloc
2084: libfdisk:      TAB: [0x2132970]: add entry 0x2132f20 [start=587776, end=125827071, size=125239296,  nested primary]
2084: libfdisk:      TAB: [0x2132970]: Before sort:
[0x2132970]: partition 0x2132d40 [partno=0, start=2048, end=585727, size=583680]
[0x2132970]: partition 0x2132e40 [partno=1, start=587774, end=125827071, size=125239298]
[0x2132970]: partition 0x2132f20 [partno=4, start=587776, end=125827071, size=125239296]
2084: libfdisk:      TAB: [0x2132970]: After sort:
[0x2132970]: partition 0x2132d40 [partno=0, start=2048, end=585727, size=583680]
[0x2132970]: partition 0x2132e40 [partno=1, start=587774, end=125827071, size=125239298]
[0x2132970]: partition 0x2132f20 [partno=4, start=587776, end=125827071, size=125239296]
2084: libfdisk:      CXT: [0x212ad50]: initialized:  last=2048, grain=2048
2084: libfdisk:      CXT: [0x212ad50]: partno=0, start=2048
2084: libfdisk:      CXT: [0x212ad50]: freespace analyze: partno=0, start=2048, end=585727
2084: libfdisk:      CXT: [0x212ad50]: partno=1, start=587774
2084: libfdisk:      CXT: [0x212ad50]: freespace analyze: partno=1, start=587774, end=125827071
2084: libfdisk:      TAB: [0x212d030]: analyze container 0x0x2132e40
2084: libfdisk:      CXT: [0x212ad50]: initialized:  last=587774, grain=2048
2084: libfdisk:      CXT: [0x212ad50]: partno=0, start=2048
2084: libfdisk:      CXT: [0x212ad50]: partno=1, start=587774
2084: libfdisk:      CXT: [0x212ad50]: partno=4, start=587776
2084: libfdisk:      CXT: [0x212ad50]: freespace container analyze: partno=4, start=587776, end=125827071
2084: libfdisk:      TAB: [0x212d030]: analyze container 0x0x2132e40 DONE [rc=0]
2084: libfdisk:      CXT: [0x212ad50]: partno=4, start=587776
2084: libfdisk:      CXT: [0x212ad50]: freespace behind last partition detected
2084: libfdisk:     PART: [0x2133000]: alloc
2084: libfdisk:      CXT: [0x212ad50]: LBA 125827072 -unchanged-
2084: libfdisk:      CXT: [0x212ad50]: LBA 314572799 -aligned-down-> 314570752 [grain=2048s]
2084: libfdisk:      CXT: [0x212ad50]: LBA 125827072 -unchanged-
2084: libfdisk:      CXT: [0x212ad50]: LBA 125827072 range:<125827072..314570752>, result: 125827072
2084: libfdisk:      TAB: [0x212d030]: add entry 0x2133000 [start=125827072, end=314572799, size=188745728, freespace  primary]
2084: libfdisk:      TAB: [0x2132970]: reset
2084: libfdisk:      TAB: [0x2132970]: remove entry 0x2132d40
2084: libfdisk:     PART: [0x2132d40]: reset
2084: libfdisk:     PART: [0x2132d40]: free
2084: libfdisk:      TAB: [0x2132970]: remove entry 0x2132e40
2084: libfdisk:     PART: [0x2132e40]: reset
2084: libfdisk:     PART: [0x2132e40]: free
2084: libfdisk:      TAB: [0x2132970]: remove entry 0x2132f20
2084: libfdisk:     PART: [0x2132f20]: reset
2084: libfdisk:     PART: [0x2132f20]: free
2084: libfdisk:      TAB: [0x2132970]: free
2084: libfdisk:      CXT: [0x212ad50]: get freespace DONE [rc=0]
2084: libfdisk:     PART: [0x21329e0]: resize: start unchanged
2084: libfdisk:     PART: [0x21329e0]: resize: enlarge to all possible
2084: libfdisk:     PART: [0x2133a30]: checking start=2048, size=583680
2084: libfdisk:     PART: [0x2132c60]: checking start=587776, size=125239296
2084: libfdisk:     PART: [0x2133000]: checking start=125827072, size=188745728
2084: libfdisk:     PART: [0x2132890]: resize: nothing usable after 587774
2084: libfdisk:     PART: [0x21329e0]: <<< resize: FAILED
2084: libfdisk:      ASK: [0x2132d40]: alloc
2084: libfdisk:      ASK: [0x2132d40]: do_ask for 'warnx'
Failed to resize partition #2.
2084: libfdisk:      ASK: [0x2132d40]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x2132d40]: reset
2084: libfdisk:      ASK: [0x2132d40]: free
2084: libfdisk:      TAB: [0x212d030]: reset
2084: libfdisk:      TAB: [0x212d030]: remove entry 0x2133a30
2084: libfdisk:     PART: [0x2133a30]: reset
2084: libfdisk:     PART: [0x2133a30]: free
2084: libfdisk:      TAB: [0x212d030]: remove entry 0x2132890
2084: libfdisk:     PART: [0x2132890]: reset
2084: libfdisk:     PART: [0x2132890]: free
2084: libfdisk:      TAB: [0x212d030]: remove entry 0x2132c60
2084: libfdisk:     PART: [0x2132c60]: reset
2084: libfdisk:     PART: [0x2132c60]: free
2084: libfdisk:      TAB: [0x212d030]: remove entry 0x2133000
2084: libfdisk:     PART: [0x2133000]: reset
2084: libfdisk:     PART: [0x2133000]: free
2084: libfdisk:      TAB: [0x212d030]: free
2084: libfdisk:      CXT: [0x212ad50]: set_partition() rc=-34
2084: libfdisk:     PART: [0x2132ac0]: reset
2084: libfdisk:     PART: [0x2132ac0]: free
2084: libfdisk:      ASK: [0x2132d40]: alloc
2084: libfdisk:      ASK: [0x2132d40]: do_ask for 'info'
/dev/sda2: Leaving.

2084: libfdisk:      ASK: [0x2132d40]: do_ask done [rc=0]
2084: libfdisk:      ASK: [0x2132d40]: reset
2084: libfdisk:      ASK: [0x2132d40]: free
```

This all worked perfectly before the rewrite so I'm wondering if I'm doing something wrong. My situation doesn't seem unusual and I don't understand why it would find nothing usable past 587774 which is the start of the extended partition I am trying to expand. Does anyone have any hints?

**Comment by carlf** [comment]
Also note that I have the same results with the most recent stable release:

```
root@av-measurement-db-psql-prod-sc-03:~/util-linux# ./sfdisk -v
lt-sfdisk from util-linux 2.30.2
```

**Comment by karelzak** [comment]
Yes, this is disadvantage of the old versions. The problem should be fixed since commit 5ec4400dec75864b475a8099ba31807fd1272af4. 

You can try the current master branch or already released v2.31-rc2 where the resize code is able to work with extended and logical partitions.
